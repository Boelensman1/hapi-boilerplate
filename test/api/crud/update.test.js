const path = require('path')
const setUpCrudTest = require('test/api/setUpCrudTest')
const { role, post } = require('models')

const seedLocation = path.join(__dirname, 'update.seed.yaml')

const { update } = require('api/crud')

describe('Test update crud function', () => {
  test('Update', async () => {
    const { request, h, seed } = await setUpCrudTest(seedLocation)

    request.payload = {
      name: 'New Name',
    }
    request.params = { id: seed.role[0].id }

    const handler = update('role', role).handler.bind(null, request, h)
    const {
      body: { result },
    } = await handler()

    expect(result).toMatchSnapshot({
      id: seed.role[0].id,
      name: request.payload.name,
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })
  })

  test('Remove virtual properties', async () => {
    const { request, h, seed } = await setUpCrudTest(seedLocation)

    request.payload = {
      title: 'New Post Title',
      titleWithAuthor: 'Virtual attribute',
    }
    request.params = { id: seed.post[0].id }

    const handler = update('post', post).handler.bind(null, request, h)
    const {
      body: { result },
    } = await handler()

    expect(result).toMatchSnapshot({
      id: seed.post[0].id,
      title: request.payload.title,
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })
  })

  // TODO: this seems to be bugged in objection
  test.skip('Add relation', async () => {
    const { request, h, seed, models } = await setUpCrudTest(seedLocation)

    request.payload = {
      name: 'New Name',
      user: [
        {
          username: 'new user',
          password: 'new password..',
        },
      ],
    }
    request.params = { id: seed.role[0].id }

    const handler = update('role', role).handler.bind(null, request, h)
    const {
      body: { result },
    } = await handler()

    const users = await models.users.query()
    expect(users).toHaveLength(1)
    expect(users[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
    })

    expect(result.users[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
    })
    delete result.users
    expect(result).toMatchSnapshot({
      id: seed.role[0].id,
      name: request.payload.name,
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })
  })

  test('Update to duplicate', async () => {
    const { request, h, seed } = await setUpCrudTest(seedLocation)

    request.payload = {
      name: 'test role dup',
    }
    request.params = { id: seed.role[0].id }

    const handler = update('role', role).handler.bind(null, request, h)
    const { body } = await handler()

    expect(body.statusCode).toBe(409)
    expect(body).toMatchSnapshot()
  })
})
