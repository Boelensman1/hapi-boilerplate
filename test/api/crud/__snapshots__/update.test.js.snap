// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test update crud function Remove virtual properties 1`] = `
Object {
  "author": "me",
  "contents": "test contents",
  "createdAt": Any<Date>,
  "id": 0,
  "title": "New Post Title",
  "titleWithAuthor": "New Post Title by me",
  "updatedAt": Any<Date>,
}
`;

exports[`Test update crud function Update 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "New Name",
  "post_create": 0,
  "post_delete": 0,
  "post_read": 0,
  "post_update": 0,
  "role_create": 0,
  "role_delete": 0,
  "role_read": 0,
  "role_update": 0,
  "updatedAt": Any<Date>,
  "user_create": 0,
  "user_delete": 0,
  "user_read": 0,
  "user_update": 0,
}
`;

exports[`Test update crud function Update to duplicate 1`] = `
Object {
  "statusCode": 409,
}
`;
